velox/functions/lib/Re2Functions.cpp

```diff
@@ -1193,120 +1453,202 @@ class PatternStringIterator {
-  size_t anyCharacterWildcardCount = 0;
-  // Total number of _ characters.
-  size_t singleCharacterWildcardCount = 0;
+  if (FOLLY_UNLIKELY(pattern.empty())) {
+    return PatternMetadata::fixed("");
+  }
 
+  std::vector<SubPatternKind> subPatternKinds;
+  std::vector<std::pair<size_t, size_t>> subPatternRanges;
   PatternStringIterator iterator{pattern, escapeChar};
 
   // Iterate through the pattern string to collect the stats for the simple
   // patterns that we can optimize.
+  std::ostringstream os;
+  SubPatternKind previousKind;
+  size_t currentSubPatternStart = 0;
+  size_t cursor = 0;
+
+  int32_t firstAnyCharsPatternIndex = -1;
   while (iterator.next()) {
-    const size_t currentStart = iterator.currentStart();
-    if (iterator.isWildcard()) {
-      if (wildcardStart == -1) {
-        wildcardStart = currentStart;
-      }
+    SubPatternKind currentKind;
+    if (iterator.isSingleCharWildcard()) {
+      currentKind = SubPatternKind::kSingleCharWildcard;
+    } else if (iterator.isAnyCharsWildcard()) {
+      currentKind = SubPatternKind::kAnyCharsWildcard;
+    } else {
+      currentKind = SubPatternKind::kLiteralString;
+    }
 
-      if (iterator.isSingleCharWildcard()) {
-        ++singleCharacterWildcardCount;
-      } else {
-        ++anyCharacterWildcardCount;
-      }
+    // New sub pattern occurs.
+    if (currentKind != previousKind && cursor > 0) {
+      subPatternKinds.push_back(previousKind);
+      subPatternRanges.push_back(
+          {currentSubPatternStart, cursor - currentSubPatternStart});
+      currentSubPatternStart = cursor;
+    }
 
-      if (!iterator.isPreviousWildcard()) {
-        ++numWildcardSequences;
-      }
+    // Advance the cursor.
+    std::string_view currentChar = iterator.current();
+    cursor += currentChar.size();
+    previousKind = currentKind;
 
-      // Mark the end of the fixed pattern.
-      if (fixedPatternStart != -1 && fixedPatternEnd == -1) {
-        fixedPatternEnd = currentStart - 1;
-      }
-    } else {
-      // Record the first fixed pattern start.
-      if (fixedPatternStart == -1) {
-        fixedPatternStart = currentStart;
-      } else {
-        // This is not the first fixed pattern, not supported, so fallback.
-        if (iterator.isPreviousWildcard()) {
-          return PatternMetadata{PatternKind::kGeneric, 0};
-        }
-      }
-    }
+    os << iterator.current();
   }
-
-  // The pattern end may not been marked if there is no wildcard char after
-  // pattern start, so we mark it here.
-  if (fixedPatternStart != -1 && fixedPatternEnd == -1) {
-    fixedPatternEnd = patternLength - 1;
+  std::string unescapedPattern = os.str();
+
+  // Handle the last sub-pattern.
+  subPatternKinds.push_back(previousKind);
+  subPatternRanges.push_back(
+      {currentSubPatternStart, cursor - currentSubPatternStart});
+
+  const auto numSubPatterns = subPatternKinds.size();
+  SubPatternStats literalStats{SubPatternKind::kLiteralString};
```

Comment:
maybe use std::array<3> here to make the code a bit more readable.

Reply:
You mean use `std::array<SubPatternStats, 3> stats` and replace `literalStats` with `stats[0]`,  replace `singleCharWildcardStats` with `stats[1]` etc?

Reply:
Something like that. stats[kind] would be more readable than hard-coded 0, 1, 2.

Reply:
Something like?

```
std::array<SubPatternStats, 3> stats;
size_t kLiteral = 0;
size_t kSingleCharWild = 1;
size_t kAnyCharWild = 2;

then use stats[kLiteral], stats[kSingleCharWild]?
```

Reply:
Or convert SubPatternKind to a normal enum?

```
enum SubPatternKind {
  /// e.g. '___'.
  kSingleCharWildcard = 0,
  // e.g. '%%'.
  kAnyCharsWildcard = 1,
  // e.g. 'abc'.
  kLiteralString = 2
};
```

Reply:
That would be nice. Just make sure to keep "enum class".

Reply:
if we keep `enum class`, we have to introduce another set of constants for the sub patterns which seems tedious, and need to maintain the relationship between these two set of enum/constants?

Reply:
Not necessarily.

```
class MemoryAllocationBenchMark {
 public:
  enum class Type {
    kMalloc = 0,
    kMmap = 1,
  };
```

Reply:
https://en.cppreference.com/w/cpp/language/enum

Reply:
Hmm... the example from the link:

```
int main()
{
    enum class Color { red, green = 20, blue };
    Color r = Color::blue;
 
    switch(r)
    {
        case Color::red  : [std::cout](http://en.cppreference.com/w/cpp/io/cout) << "red\n";   break;
        case Color::green: [std::cout](http://en.cppreference.com/w/cpp/io/cout) << "green\n"; break;
        case Color::blue : [std::cout](http://en.cppreference.com/w/cpp/io/cout) << "blue\n";  break;
    }
 
    // int n = r; // error: no implicit conversion from scoped enum to int
    int n = static_cast<int>(r); // OK, n = 21
    [std::cout](http://en.cppreference.com/w/cpp/io/cout) << n << '\n'; // prints 21
}
```

To use Color as int, we need to do a `static_cast`, you mean we do something like: `stats[static_cast<int>(subPatternKind)]`?